exonum/src/encoding/protobuf/mod.rs

```diff
@@ -0,0 +1,245 @@
+// limitations under the License.
+
+//! Protobuf generated structs and traits for conversion.
+
+// For rust-protobuf generated files.
+#![allow(bare_trait_objects)]
+#![allow(renamed_and_removed_lints)]
+
+mod blockchain;
+mod helpers;
+mod protocol;
+
+pub use self::blockchain::{Block, ConfigReference, TxLocation};
+pub use self::helpers::{BitVec, Hash, PublicKey};
+pub use self::protocol::{
+    BlockRequest, BlockResponse, Connect, PeersRequest, Precommit, Prevote, PrevotesRequest,
+    Propose, ProposeRequest, Status, TransactionsRequest, TransactionsResponse,
+};
+
+use bit_vec;
+use chrono::{DateTime, TimeZone, Utc};
+use crypto;
+use encoding::Error;
+use helpers::{Height, Round, ValidatorId};
+use messages::BinaryForm;
+use protobuf::{well_known_types, Message, RepeatedField};
+
+/// Used for establishing correspondence between rust struct
+/// and protobuf rust struct
+pub trait ToProtobuf: Sized {
+    /// Type of the protobuf clone of Self
+    type ProtoStruct: Message;
+
+    /// Struct -> ProtoStruct
+    fn to_pb(&self) -> Self::ProtoStruct;
+
+    /// ProtoStruct -> Struct
+    fn from_pb(pb: Self::ProtoStruct) -> Result<Self, ()>;
+}
+
+impl<T> BinaryForm for T
+where
+    T: ToProtobuf,
+{
+    fn encode(&self) -> Result<Vec<u8>, Error> {
+        Ok(self.to_pb().write_to_bytes().unwrap())
+    }
+
+    fn decode(buffer: &[u8]) -> Result<Self, Error> {
+        let mut pb = <Self as ToProtobuf>::ProtoStruct::new();
+        pb.merge_from_bytes(buffer).unwrap();
+        Self::from_pb(pb).map_err(|_| "Conversion from protobuf error".into())
+    }
+}
+
+/// Used for establishing correspondence between rust struct field
+/// and protobuf value
+pub trait ProtobufValue: Sized {
+    /// Type of value that is returned with pb.take_field() method
+    type ProtoValue;
+    /// RustField -> ProtobufRustField
+    fn to_pb_field(&self) -> Self::ProtoValue;
+    /// ProtobufRustField -> RustField
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()>;
+}
+
+impl<T> ProtobufValue for T
+where
+    T: ToProtobuf,
+{
+    type ProtoValue = <Self as ToProtobuf>::ProtoStruct;
+
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.to_pb()
+    }
+
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Self::from_pb(pb)
+    }
+}
+
+impl ToProtobuf for crypto::Hash {
+    type ProtoStruct = Hash;
+
+    fn to_pb(&self) -> Hash {
+        let mut hash = Hash::new();
+        hash.set_data(self.as_ref().to_vec());
+        hash
+    }
+
+    fn from_pb(pb: Hash) -> Result<Self, ()> {
+        let data = pb.get_data();
+        crypto::Hash::from_slice(data).ok_or(())
+    }
+}
+
+impl ToProtobuf for crypto::PublicKey {
+    type ProtoStruct = PublicKey;
+
+    fn to_pb(&self) -> PublicKey {
+        let mut key = PublicKey::new();
+        key.set_data(self.as_ref().to_vec());
+        key
+    }
+
+    fn from_pb(pb: PublicKey) -> Result<Self, ()> {
```

Comment:
Nit / future work: I think it may make sense to explicitly test that `from_pb` fails both for undersized and oversized data.